gateway/gateway.go

```diff
@@ -62,7 +63,16 @@ func (g *APIGateway) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	if backendService.StripPath {
 		urlPath = strings.TrimPrefix(req.URL.Path, backendService.Path)
 	} else {
-		urlPath = backendService.Path
+		urlPath = req.URL.Path
+	}
+
+	if backendService.RewriteMatch != "" && backendService.RewriteReplace != "" {
+		re, err := regexp.Compile(backendService.RewriteMatch)
```

Comment:
If I understand correctly this function is called for every single call to the gateway. Compiling the regex expression for every call is costly.
+ Could the RewriteMatch be stored as a `*regexp.Regexp`  in the backendService struct directly? 


Besides, there is a lot of reconciliation logic going on between the `clients` map and the actual `backendService`.

+ I think we could simplify this part and avoid the concurrent access latency to the client map (mutex.Lock) by directly embedding the client in the backendService struct (there seems to be a 1 to 1 relationship between clients and backendServices anyway).
    + It would also be simpler to refresh clients because we would just have to iterate over the `backendServices`instead of (once again) reconcilliating the `clients` map and the `backendServices`

Reply:
@CCOLLOT , Great points. 
Both points you've mentioned can be implemented when the backendService struct is constructed.
btw why do we even need to refresh clients? it looks like all it does is re-assigning the transport with the configuration provided, which is not needed to be refreshed, only when a configuration is modified. correct me if im wrong

Reply:
also i've found something disturbing that [this](https://github.com/Frontman-Labs/frontman/blob/main/gateway/gateway.go#L240) is blocking for the first service in the slice

Reply:
@hyperioxx

Reply:
Sorry both @amityahav @CCOLLOT, I didn't see these !

```
If I understand correctly this function is called for every single call to the gateway. Compiling the regex expression for every call is costly.
```
@CCOLLOT Yes you are correct and yes it is costly, That's a very good point and I agree with your approach !

```
btw why do we even need to refresh clients?
```
@amityahav Yeah i just looked back at that code and i don't know what the hell i was thinking at the time xD it's not even needed